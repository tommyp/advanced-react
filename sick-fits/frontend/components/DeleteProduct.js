/* eslint-disable no-alert */
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
/* eslint-disable no-restricted-globals */

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

const update = (cache, payload) => {
  cache.evict(cache.identify(payload.data.deleteProduct));
};

export default function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update,
  });
  return (
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        if (confirm('Are you sure you want to delete this item')) {
          deleteProduct().catch((err) => alert(err.mesage));
        }
      }}
    >
      {children}
    </button>
  );
}
