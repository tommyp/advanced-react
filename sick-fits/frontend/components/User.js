/* eslint-disable import/prefer-default-export */
import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';

export const CURRENT_USER_QUERY = gql`
query CURRENT_USER_QUERY {
  authenticatedItem {
    ... on User {
      id 
      email
      name
      cart {
        id
        quantity
        product {
          id
          price
          name
          description
          photo {
            image {
              publicUrlTransformed
            }
          }
        }
      }
    }
  }
}
`;

export function useUser() {
  const { data } = useQuery(CURRENT_USER_QUERY);

  return data?.authenticatedItem;
}

export const SIGN_OUT_MUTATION = gql`
mutation {
  endSession
}
`;
