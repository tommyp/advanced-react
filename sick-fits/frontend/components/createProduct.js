import useForm from "../lib/useForm";
import Form from "./styles/Form";
import gql from 'graphql-tag';
import {useMutation} from '@apollo/client'
import DisplayError from './ErrorMessage';

const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # which variables are getting passed in? And what types are they
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: { 
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE" 
        photo: {
          create: {
            image: $image
            altText: $name
          }
        }
      }
    ) {
      id
      price
      description
      name
    }
  }
`;

export default function CreateProduct() {
  const { inputs, handleChange, clearForm, resetForm } = useForm({
    image: "",
    name: "Nice shoes",
    price: 32412,
    description: "These are the best shoes",
  });

  const [createProduct, {loading, error, data}] = useMutation(CREATE_PRODUCT_MUTATION, {
    variables: inputs
  })

  return (
    <Form
      onSubmit={async (e) => {
        e.preventDefault();
        
        // submit the input fields to the backend
        const res = await createProduct()
        console.log(res)
        clearForm()
      }}
    >
      <DisplayError error={error}/>
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="image">
          Image
          <input
            required
            type="file"
            name="image"
            id="image"
            placeholder="image"
            onChange={handleChange}
          />
        </label>

        <label htmlFor="name">
          Name
          <input
            type="text"
            name="name"
            id="name"
            placeholder="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="price">
          Price
          <input
            type="number"
            name="price"
            id="price"
            placeholder="price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>

        <label htmlFor="description">
          Description
          <textarea
            name="description"
            id="description"
            placeholder="description"
            value={inputs.description}
            onChange={handleChange}
          ></textarea>
        </label>

        <button type="submit">+ Add Product</button>
      </fieldset>
    </Form>
  );
}
