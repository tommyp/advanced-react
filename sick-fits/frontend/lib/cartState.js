/* eslint-disable import/prefer-default-export */
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider!
  // We will store data (state) and functionality
  // (updaters) in here and anyone can access it via the consumer

  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  return (
    <LocalStateProvider value={{ cartOpen, toggleCart }}>
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for accessing the cart local state

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
