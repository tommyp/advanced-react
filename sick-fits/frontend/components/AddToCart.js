import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useCart } from '../lib/cartState';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
mutation ADD_TO_CART_MUTATION($id: ID!) {
  addToCart(productId: $id) {
    id
  }
}
`;

export default function AddToCart({ id }) {
  const { toggleCart } = useCart();

  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: {
      id,
    },
    refetchQueries: [{
      query: CURRENT_USER_QUERY,
    }],
  });

  const handleClick = () => {
    addToCart();
    toggleCart();
  };

  return (
    <button disabled={loading} type="button" onClick={handleClick}>
      Add
      {' '}
      {loading && 'ing'}
      {' '}
      to cart ðŸ›’
    </button>
  );
}
