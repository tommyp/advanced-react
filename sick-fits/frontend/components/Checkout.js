import styled from 'styled-components';
import { loadStripe } from '@stripe/stripe-js';
import {
  CardElement, Elements, useElements, useStripe,
} from '@stripe/react-stripe-js';
import { useState } from 'react';
import nProgress from 'nprogress';
import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import { useRouter } from 'next/dist/client/router';
import SickButton from './styles/SickButton';
import { useCart } from '../lib/cartState';
import { CURRENT_USER_QUERY } from './User';

const CheckoutFormStyles = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const ErrorStyles = styled.p`
  font-size: 2rem;
  color: var(--red);
`;

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id 
        name
      }
    }
  }
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState();
  const stripe = useStripe();
  const elements = useElements();
  const router = useRouter();
  const { toggleCart } = useCart();

  const [checkout, { error: graphQLError }] = useMutation(CREATE_ORDER_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function handleSubmit(e) {
    // 1. stop the form from submitting and turn the loader on
    e.preventDefault();
    setLoading(true);

    // 2. start the page transition
    nProgress.start();

    // 3. create the payment method via stripe (token comes back here if successful)
    const { error: err, paymentMethod } = await stripe.createPaymentMethod({
      type: 'card',
      card: elements.getElement(CardElement),
    });

    console.log(paymentMethod);

    // 4. handle any errors from stripe
    if (err) {
      setError(err);
      nProgress.done();
      return; // stops the checkout from happening
    }

    // 5. send the tokebn from step 3 to our keystone server via a custom mutation
    const order = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });

    console.log('Finished with the order');
    console.log(order);
    // 6. change the page to view the order
    router.push({
      pathname: '/order[id]',
      query: {
        id: order.data.checkout.id,
      },
    });

    // 7. close the cart
    toggleCart();
    // 8. turn the loader off
    setLoading(false);
    nProgress.done();
  }

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      {error && <ErrorStyles>{error.message}</ErrorStyles>}
      {graphQLError && <ErrorStyles>{graphQLError.message}</ErrorStyles>}
      <CardElement />
      <SickButton>
        Checkout now
      </SickButton>
    </CheckoutFormStyles>
  );
}

export default function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}
