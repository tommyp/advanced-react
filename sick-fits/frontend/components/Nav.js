import Link from 'next/link';
import { useMutation } from '@apollo/client';
import NavStyles from './styles/NavStyles';
import { useUser, SIGN_OUT_MUTATION, CURRENT_USER_QUERY } from './User';
import { useCart } from '../lib/cartState';
import CartCount from './CartCount';

export default function Nav() {
  const user = useUser();

  const [endSession] = useMutation(SIGN_OUT_MUTATION, {
    refetchQueries: [{
      query: CURRENT_USER_QUERY,
    }],
  });

  const signOut = () => {
    endSession();
  };

  const { toggleCart } = useCart();

  return (
    <NavStyles>
      <Link href="/products">Products</Link>
      {
        user && (
        <>
          <Link href="/sell">Sell</Link>
          <Link href="/orders">Orders</Link>
          <Link href="/account">Account</Link>
          <button type="button" onClick={signOut}>signout</button>
          <button type="button" onClick={toggleCart}>
            cart
            {' '}
            <CartCount count={user.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)} />
          </button>

        </>
        )
      }
      {
        !user && (
          <>
            <Link href="/signin">Sign in</Link>
          </>
        )
      }

    </NavStyles>
  );
}
