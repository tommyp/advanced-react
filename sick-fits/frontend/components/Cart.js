import styled from 'styled-components';
import CartStyles from './styles/CartStyles';
import { useUser } from './User';
import Supreme from './styles/Supreme';
import formatMoney from '../lib/formatMoney';
import calcTotalPrice from '../lib/calcTotalPrice';
import { useCart } from '../lib/cartState';
import CloseButton from './styles/CloseButton';
import RemoveFromCart from './RemoveFromCart';
import Checkout from './Checkout';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr 1fr;
  img {
    margin-right: 1rem;
  }
  h3, p {
    margin: 0;
  }
`;

const CartItem = ({ item }) => {
  const { product } = item;
  return (
    <CartItemStyles>
      <img width={100} src={product?.photo.image?.publicUrlTransformed} alt={product.name} />
      <div>
        <h3>
          {product.name}
        </h3>
        <p>
          {formatMoney(product.price * item.quantity)}
          {' '}
          -
          <em>
            {item.quantity}
            {' '}
            &times;
            {' '}
            {formatMoney(product.price)}
            {' '}
            each
          </em>
        </p>
      </div>
      <RemoveFromCart id={item.id} />

    </CartItemStyles>
  );
};

export default function Cart() {
  const user = useUser();
  const { cartOpen, toggleCart } = useCart();

  if (!user) return null;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>
          {user.name}
          &apos;s Cart
        </Supreme>
        <CloseButton type="button" onClick={toggleCart}>&times;</CloseButton>
      </header>

      <ul>
        {user.cart.map((item) => (
          <CartItem key={item.id} item={item} />
        ))}
      </ul>

      <footer>
        <p>
          {formatMoney(calcTotalPrice(user.cart))}
        </p>
        <Checkout />
      </footer>
    </CartStyles>
  );
}
