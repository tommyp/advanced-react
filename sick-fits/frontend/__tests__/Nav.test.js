import { MockedProvider } from '@apollo/react-testing';
import { render, screen } from '@testing-library/react';

import Nav from '../components/Nav';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';
import { CartStateProvider } from '../lib/cartState';

// mocks for signed out, signed in and signed in with cart items

const noSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: null } },
  },
];

const signedInMocks = [{
  request: { query: CURRENT_USER_QUERY },
  result: { data: { authenticatedItem: fakeUser() } },
}];

const signedInMocksWithCartItems = [{
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      authenticatedItem: fakeUser({
        cart: [fakeCartItem()],
      }),
    },
  },
}];

describe('<Nav/>', () => {
  it('renders signed out user', () => {
    const { container } = render(
      <MockedProvider mocks={noSignedInMocks}>
        <CartStateProvider>
          <Nav />
        </CartStateProvider>
      </MockedProvider>,
    );

    expect(container).toHaveTextContent('Sign in');
    expect(container).toMatchSnapshot();
  });

  it('renders a full nav when signed in', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={signedInMocks}>
        <CartStateProvider>
          <Nav />
        </CartStateProvider>
      </MockedProvider>,
    );

    await screen.findByText('Account');

    expect(container).not.toHaveTextContent('Sign in');
    expect(container).toHaveTextContent('Sign out');
    expect(container).toHaveTextContent('My cart');
    expect(container).toMatchSnapshot();
  });

  it('renders the amount of items in the cart', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={signedInMocksWithCartItems}>
        <CartStateProvider>
          <Nav />
        </CartStateProvider>
      </MockedProvider>,
    );

    await screen.findByText('Account');

    expect(container).not.toHaveTextContent('Sign in');
    expect(container).toHaveTextContent('Sign out');
    expect(container).toHaveTextContent('My cart 3');

    expect(container).toMatchSnapshot();
  });
});
