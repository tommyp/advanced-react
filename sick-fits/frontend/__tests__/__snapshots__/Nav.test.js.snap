// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a full nav when signed in 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 cHJqBj"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <button
      type="button"
    >
      Sign out
    </button>
    <button
      type="button"
    >
      My cart
       
      <span
        class="CartCount__AnimationStyles-sc-xxvp4g-1 hNdwpB"
      >
        <div>
          <div
            class="CartCount__Dot-sc-xxvp4g-0 iephzA count"
          >
            0
          </div>
        </div>
      </span>
    </button>
  </ul>
</div>
`;

exports[`<Nav/> renders signed out user 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 cHJqBj"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/signin"
    >
      Sign in
    </a>
  </ul>
</div>
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<div>
  <ul
    class="NavStyles-sc-11c0d2g-0 cHJqBj"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <button
      type="button"
    >
      Sign out
    </button>
    <button
      type="button"
    >
      My cart
       
      <span
        class="CartCount__AnimationStyles-sc-xxvp4g-1 hNdwpB"
      >
        <div>
          <div
            class="CartCount__Dot-sc-xxvp4g-0 iephzA count"
          >
            3
          </div>
        </div>
      </span>
    </button>
  </ul>
</div>
`;
